TFLearn Estimator Summary Writer Fix

thesuperzapper opened the pull request:
Fixes #7554

thesuperzapper committed:
Merge remote-tracking branch 'refs/remotes/tensorflow/master'

thesuperzapper committed:
Fix evaluation summary writing. TFLearn

tensorflow-jenkins commented:
Can one of the admins verify this patch?

martinwicke commented on tensorflow/contrib/learn/python/learn/estimators/estimator.py:
```diff
@@ -321,11 +321,15 @@ def _write_dict_to_summary(output_dir,
-    if (isinstance(dictionary[key], np.float32) or
-        isinstance(dictionary[key], float)):
+    if (isinstance(dictionary[key], np.float32) or 
+            isinstance(dictionary[key], float)):
```
Can you remove the extra whitespace here?

thesuperzapper replied:
Yep

martinwicke commented on tensorflow/contrib/learn/python/learn/estimators/estimator.py:
```diff
@@ -321,11 +321,15 @@ def _write_dict_to_summary(output_dir,
       value.simple_value = float(dictionary[key])
+    elif (isinstance(dictionary[key], np.int64) or
+              isinstance(dictionary[key], np.int32) or
+              isinstance(dictionary[key], int)):
```
Is int even different from both of these?

martinwicke commented on tensorflow/contrib/learn/python/learn/estimators/estimator.py:
```diff
@@ -321,11 +321,15 @@ def _write_dict_to_summary(output_dir,
+      value.simple_value = int(dictionary[key])
     else:
-      logging.warn('Skipping summary for %s, must be a float or np.float32.',
+      logging.warn('Skipping summary for %s, must be a float, np.float32, np.int64, np.int32 or int.',
```
Here also, I don't think int can ever be neither int32, nor int64.

martinwicke reviewed (changes requested):
Thank you for this! Can you please add a test that exercises this new code?

thesuperzapper commented:
Sorry for disappearing for a while, been super busy, but we really need to get this fixed. We need the int type because people can write their own MetricSpec's.

thesuperzapper committed:
Update estimator.py

Fixed White Space

dandelionmane commented:
@thesuperzapper Please add a test as martinwicke requested.

drpngx commented:
@thesuperzapper feel free to add tests. Thanks!

thesuperzapper commented:
I am not 100% sure how to write test cases.

martinwicke commented:
You can look at the estimator_test.py file to see other tests.
We should have test cases that exercise these new code paths.

I'm a bit unclear what failure this addresses. I think the change is fine in general though.

Can you make the same change (and test) to tensorflow/python/estimators/estimator.py as well? The _write_dict_to_summary code looks the same.

martinwicke commented:
@thesuperzapper Ping?

thesuperzapper committed:
Merge pull request #1 from thesuperzapper/tflearn-evaluation-bug

Tflearn evaluation bug

thesuperzapper committed:
Merge pull request #2 from tensorflow/master

Update to master

thesuperzapper committed:
Merge pull request #3 from thesuperzapper/master

Update to master

thesuperzapper committed:
Fix similar issue. 

Fix similar issue under, `tensorflow/python/estimators/estimator.py`

thesuperzapper commented:
_Note: I added the same fix to `tensorflow/python/estimators/estimator.py`._

This pull request simply allows us to write int-type values in the summary file.

A side effect of this pull request will be **writing the value of 'global_step' at 'global_step'** -- a useless piece of information -- this is because 'global_step' is inside the ['eval_results'](https://github.com/tensorflow/tensorflow/blob/1a1527ab1fcffd94f692b1301fdbe87903ce7bdb/tensorflow/contrib/learn/python/learn/estimators/estimator.py#L841) & ['dictionary'](https://github.com/tensorflow/tensorflow/blob/1a1527ab1fcffd94f692b1301fdbe87903ce7bdb/tensorflow/contrib/learn/python/learn/estimators/estimator.py#L330) objects, (Which would not previously have been written by [`_write_dict_to_summary`](https://github.com/tensorflow/tensorflow/blob/1a1527ab1fcffd94f692b1301fdbe87903ce7bdb/tensorflow/contrib/learn/python/learn/estimators/estimator.py#L316) because 'global_step' is an int type, as seen in #7554)

Therefore, we probably should remove the 'global_step' key from the 'eval_results' dictionary. (Or filter it in `_write_dict_to_summary`), otherwise 'global_step' by 'global_step' is displayed in TensorBoard. 

But also note that we return the 'eval_results' dictionary in [ `_evaluate_model`](https://github.com/tensorflow/tensorflow/blob/1a1527ab1fcffd94f692b1301fdbe87903ce7bdb/tensorflow/contrib/learn/python/learn/estimators/estimator.py#L853). Therefore, I would like some advice on the best way of removing/filtering the 'global_step' key, so that we don't break this return, or refactor stuff unnecessarily.

martinwicke commented:
I think the right solution here is to filter global_step out inside _write_dict_to_summary. Nobody needs to see that graph, it's boring.

I still want a test. Without one, I guarantee that this will break sooner or later.

vrv commented:
another ping for @thesuperzapper !

thesuperzapper commented:
@vrv 
Sorry guys, I have been stupidly busy! I am still not sure how the testing framwork works.

Do I just add a function to the [estimator_test.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/estimators/estimator_test.py) file?

Also, do you think hard-coded filtering the `global_step` key inside [_write_dict_to_summary](https://github.com/tensorflow/tensorflow/blob/1a1527ab1fcffd94f692b1301fdbe87903ce7bdb/tensorflow/contrib/learn/python/learn/estimators/estimator.py#L331) is good enough?

vrv commented:
Yes, a test to estimator_test.py would be the right place.

(I don't know about the 'global step' key filtering, punting to @martinwicke and @ispirmustafa )

martinwicke commented:
You can use the `tf.train.get_global_step` method to check whether a value you're about to write is the global step tensor.

ispirmustafa commented:
+1 for "filter global_step out inside _write_dict_to_summary"
You can do if before calling the _write_dict_to_summary.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

ispirmustafa reviewed (changes requested):
global step should be filtered out before calling this function.

drpngx commented:
@thesuperzapper any luck on this?

thesuperzapper commented:
I will look at this later today, but it's pretty much just making the current changes, and then stopping the writer from writing the global step.

thesuperzapper committed:
Filter out "global_step"

This is seemingly the best way to ensure we don't write "global_step" to the summary, (which would be useless), while not modifying the dictionary. 

(Pull request for the other `estimator.py` incoming.)

thesuperzapper committed:
Filter "global_step" [part 2]

Same as other `estimator.py`.

rmlarsen commented:
@ispirmustafa @martinwicke could you take another look, please?

rmlarsen commented:
@tensorflow-jenkins test this please
